Layer preview stacked

mjg opened the pull request:
Here is the follow-up to PR #2674.

Things to note in addition to the commit messages:

- The existing layer preview shows layers irrespective of the "active" check-box (which is useful); I kept this behaviour.
- The new sidebar icon uses some screen real estate that - until a few days ago - was unused anyways. I still think both the sidebar "page" (tab) icon and the close box belong there.
- The technical implementation makes it easy to use a checkbox instead of a new sidebar tab.

rolandlo commented:
This looks useful and seems to work fine. Some comments:
- When the sidebar is made narrow, there is a dropdown menu for selecting the sidebar tab. In this dropdown menu I currently see the string "LayerPreview" duplicated, although it is for two different tabs. That's confusing.
- The new icon clashes with #1735. Eventually we will need consistency.
- There should definitively be no `inkscape:export-filename=` (personal path names) in the svg-file data.
- Icons will need to be contributed under the GPLv2 license.

mjg commented:
> * When the sidebar is made narrow, there is a dropdown menu for selecting the sidebar tab. In this dropdown menu I currently see the string "LayerPreview" duplicated, although it is for two different tabs. That's confusing.

Good catch. Changed now.

> * The new icon clashes with #1735. Eventually we will need consistency.

I went by similarity to the existing one, since layer and layerstack are similar views to the same data.

> * There should definitively be no `inkscape:export-filename=` (personal path names) in the svg-file data.

That's in many places. I inserted a commit cleaning up all of them now.

> * Icons will need to be contributed under the GPLv2 license.

How do I do that? Also: Since I merely added an arrow to the existing icon, am I entitled to change the license?

To be clear: I contribute under the existing project license, which is GPLv2, so no problem on my side. Should I specify that with a sign-off or such?

Technius commented:
@rolandlo I think we should document the licenses of each contributed file in a `COPYRIGHT.md` or something as recommended in #2148.

mjg commented:
I checked the icons redesign PR now. Do you prefer me to drop the icon renaming (and clean-up) commits here?

rolandlo commented:
> I checked the icons redesign PR now. Do you prefer me to drop the icon renaming (and clean-up) commits here? [In fact: it does some renaming, too.]

Thinking about it, yes I would prefer that because like that it will be easier to get the icons redesign PR #1735 finished once your PR is merged. You could leave a note in #1735 what renaming you suggest.

About the licensing: We will have to redo the new icon in PR #1735. So you don't need to bother about it here.

mjg commented:
I dropped all the renaming now and rebased onto current master.

Also, I name the new-icon `sidebar-layerstack` now which should be in line with the the renaming `layer -> sidebar-layer` over at #1735 for the existing icon which I copy from for now. (I omitted the homedir from the copy.)

rolandlo reviewed (approved):
A minor nitpick:
There are some magical numbers in `PreviewJob.cpp`. It may be better to define some `constexpr` to reveal their meaning. 
Otherwise everything looks fine to me. So either change that or, if you disagree, tell me to merge.

mjg commented:
> A minor nitpick:
> There are some magical numbers in `PreviewJob.cpp`. It may be better to define some `constexpr` to reveal their meaning.
> Otherwise everything looks fine to me. So either change that or, if you disagree, tell me to merge.

I completely agree that this is all magic :)

"-1" (background layer) and "-100" (all layers) where there before. I added "-layer-1000" to denote stacked up to "layer", which arguable passes the fine line between "magic" and "weird"... I could make this more legible by introducing a mode parameter to `PreviewJob::drawPage()` which takes an enum or const argument - in fact, I could reuse or refactor the `PreviewRenderType` from `SidebarPreviewBaseEntry.h`. I have to check which includes go where first.

mjg commented:
So, hopefully nice and clean now :)
I refactored in a separate commit (because of the amount of changes, and to check validity) and rebased the feature change on top of that (and everything on current master).

rolandlo reviewed (approved):
Thanks for cleaning this up! Looks much better now.

Testing out your PR I found a bug when switching documents with multiple annotated layers. It is not related to your PR though. I have reproduced it on 1.0.19 as well and will submit a bug report.

mjg commented on src/control/jobs/PreviewJob.cpp:
```diff
@@ -60,33 +60,76 @@ void PreviewJob::drawBackgroundPdf(Document* doc) {
+        case RENDER_TYPE_PAGE_LAYER:
+            if (layer != -1) {
+                break;  // out
+            }           // else fall through for layer == -1
+
```
Ah, cool, learned something new about c++. (c guy here ...) So, c++17 is OK, right? I'll rebase with that annotation then.

rolandlo replied:
Yes, C++17 is OK.

rolandlo commented:
Ready for merge?

rolandlo commented:
Thanks a lot. Nice feature!

ericschdt commented:
> Thanks a lot. Nice feature!

I second that! Any chance this will go into 1.1.0? :)

rolandlo commented:
> > Thanks a lot. Nice feature!
> 
> I second that! Any chance this will go into 1.1.0? :)

Sure, it will go into 1.1.0. It is already in [1.1.0+dev](https://github.com/xournalpp/xournalpp/releases/tag/nightly)
